core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/scheduling/ExpansionGreedyScheduleGenerator.scala

```diff
@@ -152,35 +160,207 @@ class ExpansionGreedyScheduleGenerator(
     Left(regionDAG)
   }
 
-  private def handleDependentLinks(
+  /**
+    * A dependee input port is one that is depended on by another input port of the same operator.
+    * The incoming edge of a dependee input port is called a dependee edge.
+    * Similarly, the other port of this dependency relationship is called a depender input port and connects
+    * to a depender edge.
+    *
+    * Core design: a dependee edge needs to be materialized, and is mapped to a region edge in the region DAG.
+    * Note: currently we assume there CANNOT be dependencies between two dependee input ports.
+    * This method reasons about the input port dependencies of a given operator during the greedy expansion-based
+    * construction of a region DAG.
+    *
+    * This method first reasons about the dependencies of the input ports of the given operator to find
+    * pairs of dependency relationships, and then enforces the dependency of each pair:
+    * All the incoming edges of a dependee port will be added to the partial region DAG as a region edge.
+    * If adding a dependee edge results in a cycle that breaks the region DAG, we use a heurestic which is to
+    * return the other depender edge and indicate that this depender edge needs to be materialized. This will
+    * break the cycle and maintain the acyclicity of the region DAG.
+    *
+    * Previously we relied purely on edges and cache read operators for implementing materializations for
+    * materialized edges and find regions in this method.
+    *
+    * After introducing materailizations on output and input ports, materializing an
+    * edge could result in an operator that does not have any edges connected to one or more of its input
+    * ports (i.e., it becomes a "starter" operator in a region). For such input ports, we can only use port to
+    * find regions.
+    *
+    * @param physicalOpId The id of the input physical operator on which we need to handle input port dependies.
+    * @param regionDAG The partial region DAG that is always acyclic.
+    * @return Optionally a set of [[PhysicalLink]]s to do materialization-replacements on.
+    */
+  private def handleInputPortDependencies(
       physicalOpId: PhysicalOpIdentity,
       regionDAG: DirectedAcyclicGraph[Region, RegionLink]
   ): Option[Set[PhysicalLink]] = {
-    // for operators like HashJoin that have an order among their blocking and pipelined inputs
+    // for operators like HashJoin's Probe that have dependencies between their input ports
```

Comment:
let's be consistent. the function name is `.getInputPortsInProcessingOrder`, while the comment says `dependencies`. I suggest changing either side to align them.